fetcher/src/job/job.utils.ts

```diff
@@ -2,40 +2,93 @@ import axios from 'axios'
+import { Logger } from '@nestjs/common'
 
 export function buildUrl(host: string, path: string) {
   const url = [host, path].join('/')
   return url.replace(/([^:]\/)\/+/g, '$1')
 }
 
+function isProxyDefined(adapter) {
+  return (
+    adapter.proxy !== undefined &&
+    adapter.proxy.protocol !== undefined &&
+    adapter.proxy.host !== undefined &&
+    adapter.proxy.port !== undefined
+  )
+}
+
+const INVALID_DATA = -1
+async function fetchCall(url: string, options, logger) {
+  try {
+    // const result = (await axios.get('https://httpbin.org/ip', options)).data
+    return (await axios.get(url, options)).data
+  } catch (e) {
+    logger.error(`Error in fetching data from ${url}: ${e.message}`)
+    logger.error(e)
+    return INVALID_DATA
+  }
+}
+
+async function fetchRawDataWithProxy(adapter, logger) {
+  return fetchCall(
+    adapter.url,
+    {
+      method: adapter.method,
+      headers: adapter.headers,
+      proxy: {
+        protocol: adapter.proxy.protocol,
+        host: adapter.proxy.host,
+        port: adapter.proxy.port
+      }
+    },
+    logger
+  )
+}
+
+async function fetchRawDataWithoutProxy(adapter, logger) {
+  return fetchCall(
+    adapter.url,
+    {
+      method: adapter.method,
+      headers: adapter.headers
+    },
+    logger
+  )
+}
+
 /**
  * Fetch data from data sources defined in `adapter`.
  *
  * @param {} adapter Single data adapter to define which data to fetch.
  * @param {} NestJs logger
  * @return {number} aggregatedresults
  */
-export async function fetchData(adapter, logger) {
+export async function fetchData(adapterList, logger) {
   const data = await Promise.allSettled(
-    adapter.map(async (a) => {
-      const options = {
-        method: a.method,
-        headers: a.headers
-      }
-
+    adapterList.map(async (adapter) => {
       try {
-        const rawDatum = (await axios.get(a.url, options)).data
+        let rawDatum = INVALID_DATA
+        if (isProxyDefined(adapter)) {
+          rawDatum = await fetchRawDataWithProxy(adapter, logger)
+        }
+        if (rawDatum === INVALID_DATA) {
+          rawDatum = await fetchRawDataWithoutProxy(adapter, logger)
+          if (rawDatum === INVALID_DATA) {
+            throw new Error('Error in fetching data')
+          }
+        }
 
         // FIXME Build reducers just once and use. Currently, can't
         // be passed to queue, therefore has to be recreated before
         // every fetch.
-        const reducers = buildReducer(DATA_FEED_REDUCER_MAPPING, a.reducers)
+        const reducers = buildReducer(DATA_FEED_REDUCER_MAPPING, adapter.reducers)
         const datum = pipe(...reducers)(rawDatum)
         checkDataFormat(datum)
-        return { id: a.id, value: datum }
+        return { id: adapter.id, value: datum }
       } catch (e) {
-        logger.error(`Error in ${a.name}`)
+        logger.error(`Fetching with proxy ${adapter.proxy.host} failed in ${adapter.name}`)
```

Comment:
I think this error text is not correct since the error cannot come only from proxy call. In both cases (`fetchRawDataWithProxy` and `fetchRawDataWithoutProxy`) we catch the errors internally and return `INVALID_DATA` if that happened. The error coming in this `catch` branch would be either coming from line 79 (non proxy call) or from post processing code from lines 86 to 89. It would be actually better if we split it to two `try-catch`.

```typescript
export async function fetchData(adapterList, logger) {
    const data = await Promise.allSettled(
	adapterList.map(async (adapter) => {

	    let rawDatum = INVALID_DATA
	    if (isProxyDefined(adapter)) {
		rawDatum = await fetchRawDataWithProxy(adapter, logger)
	    }
	    if (rawDatum === INVALID_DATA) {
		rawDatum = await fetchRawDataWithoutProxy(adapter, logger)
		if (rawDatum === INVALID_DATA) {
		    throw new Error('Error in fetching data')
		}
	    }

	    try {
		// FIXME Build reducers just once and use. Currently, can't
		// be passed to queue, therefore has to be recreated before
		// every fetch.
		const reducers = buildReducer(DATA_FEED_REDUCER_MAPPING, adapter.reducers)
		const datum = pipe(...reducers)(rawDatum)
		checkDataFormat(datum)
		return { id: adapter.id, value: datum }
	    } catch (e) {
		logger.error(`Postprocessing of data from ${adapter.name} failed`)
	        logger.error(e)
		throw e
	    }

	})
    )

    return data.flatMap((D) => (D.status == 'fulfilled' ? [D.value] : []))
}
```